### PR TITLE
feat(TV UI): Accounts PIN login support

### PR DESCRIPTION
**Pull request by KingLucius**
- Default method on TV, If sync api supports authentication via PIN.
- You can local authenticating also.
- QR scanning for URL opening.

**Note**: CS logo can be added to QR code but the code was overkill as QR Lib recieve the logo as ByteArray so I removed it.

### CODE REVIEW

**Comment by IndusAryan** [comment]
> * Default method on TV, If sync api supports authentication via PIN.
> 
>     * You can local authenticating also.
> 
>     * QR scanning for URL opening.
> 
> 
> **Note**: CS logo can be added to QR code but the code was overkill as QR Lib recieve the logo as ByteArray so I removed it.

can we embed an image overflow on the qr image itself in centre

**Comment by KingLucius** [comment]
> > * Default method on TV, If sync api supports authentication via PIN.
> >   
> >   * You can local authenticating also.
> >   * QR scanning for URL opening.
> > 
> > **Note**: CS logo can be added to QR code but the code was overkill as QR Lib recieve the logo as ByteArray so I removed it.
> 
> can we embed an image overflow on the qr image itself in centre

it's already supported, only need the logo to be passed as byteArray, I have succesfully did it but then removed the code, I can do it again if you guys sees it's better

File: app/src/main/java/com/lagradost/cloudstream3/ui/settings/SettingsAccount.kt [hunk 1]
```diff
@@ -133,7 +142,85 @@ class SettingsAccount : PreferenceFragmentCompat(), BiometricAuthenticator.Biome
...
+                            }
+                            val dialog = builder.create()
+
+                            ioSafe {
+                                try {
+                                    val pinCodeData = api.getDevicePin()
+                                    if (pinCodeData == null) {
+                                        activity.runOnUiThread {
+                                            showToast(
+                                                activity.getString(R.string.device_pin_error_message)
```
Comment (fire-light42) [inline app/src/main/java/com/lagradost/cloudstream3/ui/settings/SettingsAccount.kt:171]: 1. toast is thread safe
2. toast can take a stringres

File: app/src/main/java/com/lagradost/cloudstream3/ui/settings/SettingsAccount.kt [hunk 2]
```diff
@@ -133,7 +142,85 @@ class SettingsAccount : PreferenceFragmentCompat(), BiometricAuthenticator.Biome
...
+                                            activity.getString(R.string.device_pin_expired_message)
+                                        )
+                                        dialog.dismissSafe(activity)
+                                    }
+                                } catch (e: Exception) {
+                                    logError(e)
+                                }
+                            }
+                        } else {
+                            api.authenticate(activity)
```
Comment (fire-light42) [inline app/src/main/java/com/lagradost/cloudstream3/ui/settings/SettingsAccount.kt:241]: this is really weird to read, reverse this if statment so this is on top and else is the rest of the code

File: app/src/main/java/com/lagradost/cloudstream3/ui/settings/SettingsAccount.kt [hunk 3]
```diff
@@ -133,7 +142,85 @@ class SettingsAccount : PreferenceFragmentCompat(), BiometricAuthenticator.Biome
...
+                                                        activity.getString(R.string.authenticated_user)
+                                                            .format(
+                                                                api.name
+                                                            )
+                                                    )
+                                                dialog.dismissSafe(activity)
+                                            }
+                                            return@ioSafe
+                                        }
+                                        delay(1000)
```
Comment (fire-light42) [inline app/src/main/java/com/lagradost/cloudstream3/ui/settings/SettingsAccount.kt:227]: this entire setup is weird and prone to errors. Just use something like this https://developer.android.com/reference/android/os/CountDownTimer instead

**Review by fire-light42 — CHANGES_REQUESTED** [review]
This is my initial review, more might come after I have tested it

File: app/src/main/res/values/strings.xml [hunk 4]
```diff
@@ -775,4 +777,8 @@
     <string name="custom_media_singluar">Media</string>
     <string name="reset_btn">Reset</string>
     <string name="cs3wiki">CloudStream Wiki</string>
+    <string name="device_pin_url_message">Visit <b>%s</b> on your smartphone or computer and enter the above code</string>
+    <string name="device_pin_error_message">"Can't get the device PIN code, try local authentication"</string>
```
Comment (fire-light42) [inline app/src/main/res/values/strings.xml:781]: Why prefix and suffix with " ????

Comment (KingLucius) [reply app/src/main/res/values/strings.xml]: Mistake

**Review by fire-light42 — APPROVED** [review]
Works, and looks very nice 👍

**Comment by KingLucius** [comment]
> Works, and looks very nice 👍 

Thanks, merged in the right time
I will add this to trakt sync api also 👍

**Pull request merged**